sdk/iot/provisioning/src/az_iot_provisioning_client.c

```diff
@@ -47,161 +54,420 @@ AZ_NODISCARD az_result az_iot_provisioning_client_init(
+  remainder = az_span_copy_u8(remainder, '#');
+  remainder = az_span_copy_u8(remainder, '\0');
 
-  *out_mqtt_topic_filter = az_span_slice(mqtt_topic_filter, 0, required_length);
+  if (out_mqtt_topic_filter_length)
+  {
+    *out_mqtt_topic_filter_length = (size_t)required_length;
+  }
 
   return AZ_OK;
 }
 
-// topic: $dps/registrations/res/202/?$rid=1&retry-after=3
-// payload:
-// {"operationId":"4.d0a671905ea5b2c8.42d78160-4c78-479e-8be7-61d5e55dac0d","status":"assigning"}
-AZ_NODISCARD az_result az_iot_provisioning_client_received_topic_payload_parse(
+AZ_INLINE az_iot_provisioning_client_registration_state registration_state_default()
+{
+  return (az_iot_provisioning_client_registration_state){ .assigned_hub_hostname = AZ_SPAN_NULL,
+                                                          .device_id = AZ_SPAN_NULL,
+                                                          .status = AZ_IOT_STATUS_OK,
+                                                          .extended_error_code = 0,
+                                                          .error_message = AZ_SPAN_NULL,
+                                                          .error_tracking_id = AZ_SPAN_NULL,
+                                                          .error_timestamp = AZ_SPAN_NULL };
+}
+
+/*
+Documented at
+https://docs.microsoft.com/en-us/rest/api/iot-dps/runtimeregistration/registerdevice#deviceregistrationresult
+  "registrationState":{
+    "x509":{},
+    "registrationId":"paho-sample-device1",
+    "createdDateTimeUtc":"2020-04-10T03:11:13.0276997Z",
+    "assignedHub":"contoso.azure-devices.net",
+    "deviceId":"paho-sample-device1",
+    "status":"assigned",
+    "substatus":"initialAssignment",
+    "lastUpdatedDateTimeUtc":"2020-04-10T03:11:13.2096201Z",
+    "etag":"IjYxMDA4ZDQ2LTAwMDAtMDEwMC0wMDAwLTVlOGZlM2QxMDAwMCI="}}
+*/
+AZ_INLINE az_result az_iot_provisioning_client_payload_error_code_parse(
+    az_json_token_member* tm,
+    az_iot_provisioning_client_registration_state* out_state)
+{
+  if (az_span_is_content_equal(AZ_SPAN_FROM_STR("errorCode"), tm->name))
+  {
+    double value;
+    AZ_RETURN_IF_FAILED(az_json_token_get_number(&tm->token, &value));
+    out_state->extended_error_code = (uint32_t)value;
+    out_state->status = (uint32_t)value / 1000;
+
+    return AZ_OK;
+  }
+
+  return AZ_ERROR_ITEM_NOT_FOUND;
+}
+
+AZ_INLINE az_result az_iot_provisioning_client_payload_registration_state_parse(
+    az_json_parser* jp,
+    az_json_token_member* tm,
+    az_iot_provisioning_client_registration_state* out_state)
+{
+  if (tm->token.kind != AZ_JSON_TOKEN_OBJECT_START)
+  {
+    return AZ_ERROR_PARSER_UNEXPECTED_CHAR;
+  }
+
+  bool found_assigned_hub = false;
+  bool found_device_id = false;
+
+  while ((!(found_device_id && found_assigned_hub))
+         && az_succeeded(az_json_parser_parse_token_member(jp, tm)))
+  {
+    if (az_span_is_content_equal(AZ_SPAN_FROM_STR("assignedHub"), tm->name))
```

Comment:
>AZ_SPAN_FROM_STR("assignedHub") [](start = 33, length = 31)

how about making all these const data as static?

Reply:
Not sure what you mean since this is already const and can only be used within this function.
Assuming you meant defining this at the top of the file as static const I don't agree: currently this is scoped only to this function which makes the code readable and harder to use in the wrong place.